core/modules/utils/utils.js

```diff
@@ -1005,4 +1005,100 @@ exports.makeCompareFunction = function(type,options) {
 	return (types[type] || types[options.defaultType] || types.number);
 };
 
+// The character that will substitute for a space in the URL
+var SPACE_SUBSTITUTE = "_";
+
+// The character added to the end to avoid ending with `.`, `?`, `!` or the like
+var TRAILER = "_";
+
+// The character that will separate out the list elements in the URL
+var CONJUNCTION = ";";
+
+// Those of the allowed url characters claimed by TW
+var CLAIMED = [SPACE_SUBSTITUTE, ":", CONJUNCTION];
+
+// Non-alphanumeric characters allowed in a URL fragment
+// More information at https://www.rfc-editor.org/rfc/rfc3986#appendix-A
+var VALID_IN_URL_FRAGMENT = "-._~!$&'()*+,;=:@/?".split("");
+
+// The subset of the pchars we will not percent-encode in permalinks/permaviews
+var SUBSTITUTES = []
+$tw.utils.each (VALID_IN_URL_FRAGMENT, function(c) {if(CLAIMED.indexOf(c) === -1) {SUBSTITUTES.push(c)}});
+
+// A regex to match the percent-encoded characters we will want to replace.
+// Something similar to the following, depending on SPACE and CONJUNCTION
+//     /(%2D|%2E|%7E|%21|%24|%26|%27|%28|%29|%2A|%2B|%3B|%3D|%40|%2F|%3F)/g
+
+var CHAR_MATCH_STR = []
+$tw.utils.each(SUBSTITUTES, function(c) {
+	CHAR_MATCH_STR.push("%" + c.charCodeAt(0).toString(16).toUpperCase())
+})
+var CHAR_MATCH = new RegExp("(" + CHAR_MATCH_STR.join("|") + ")", "g");
+
+// A regex to match the SPACE_SUBSTITUTE character
+var SPACE_MATCH = new RegExp("(\\" + SPACE_SUBSTITUTE + ")", "g");
+
+// A regex to match URLs ending with sentence-ending punctuation
+var SENTENCE_ENDING = new RegExp("(\\.|\\!|\\?|\\" + TRAILER + ")$", "g");
+
+// A regex to match URLs ending with sentence-ending punctuation plus the TRAILER
+var SENTENCE_TRAILING = new RegExp("(\\.|\\!|\\?|\\" + TRAILER + ")\\" + TRAILER + "$", "g");
+
+// An object mapping the percent encodings back to their source characters
+var PCT_CHAR_MAP = SUBSTITUTES.reduce(function (a, c) {
+	a["%" + c.charCodeAt(0).toString(16).toUpperCase()] = c
+	return a
+}, {});
+
+// Convert a URI List Component encoded string (with the `SPACE_SUBSTITUTE`
+// value as an allowed replacement for the space character) to a string
+exports.decodeTWURIList = function(s) {
+	var parts = s.replace(SENTENCE_TRAILING, "$1").split(CONJUNCTION);
+	var withSpaces = []
+	$tw.utils.each(parts, function(s) {withSpaces.push(s.replace(SPACE_MATCH, " "))});
+	var withBrackets  = []
+	$tw.utils.each(withSpaces, function(s) {withBrackets .push(s.indexOf(" ") >= 0 ? "[[" + s + "]]" : s)});
```

Comment:
please make the code readable

Reply:
Can you explain what that would entail?  I feel like I've been slowly making the code less-and-less readable to meet the criteria here.  Should I undo that, or move to a still more imperative style?

To me, this version is *much* more readable:

```js
	const withSpaces = parts.map(s => s.replace(SPACE_MATCH, " "))
	const withBrackets = withSpaces.map(s => s.indexOf(" ") > = 0 ? "[[" + s + "]]" : s)
```

and this a step further toward readablity:

```js
	const withBrackets = parts
		.map(s => s.replace(SPACE_MATCH, " "))
		.map(s => s.indexOf(" ") > = 0 ? "[[" + s + "]]" : s)
```

But I'm pretty sure that's not what you want.

Reply:
If you globally search for `$tw.utils.each(` in the TW source you can see how it's ususally used. For new code we try to stay consistent. 

```
$tw.utils.each(withSpaces,function(s){
	withBrackets.push(s.indexOf(" ") >= 0 ? "[[" + s + "]]" : s);
});
```

Reply:
Updated to match this.  I'd argue the point about readability, but this is not the time or place!  😉